Refactor notification's creation code

saraycp opened the pull request:
After some refactoring made on the Notifications code and database structure, some data had inconsistent values. So we are going to add some data migration to fix it.

This PR includes some refactoring we had to do in order to make it possible to create notifications from a data migration file. We also took the opportunity to give better names to some methods and improve the code a bit.

Having this refactoring, the next step will be adding the data migration, what will be done in a different PR.

Co-authored-by: Eduardo Navarro <enavarro@suse.com>
Co-authored-by: David Kang <dkang@suse.com>

danidoni commented on src/api/app/models/comment.rb:
```diff
@@ -49,17 +47,27 @@ def create_notification(params = {})
+
+  private
+
+  def create_notification(params = {})
```
This is creating events, not notifications, right?

saraycp replied:
Done!

danidoni commented on src/api/app/models/notification/rss_feed_item.rb:
```diff
@@ -17,6 +17,37 @@ def self.cleanup
+    notification_params.merge!(notification_dynamic_params(event))
+  end
+
+  def self.notification_dynamic_params(event)
```
What about moving this into some of the `Event` classes? I feel we're asking too much information about the event here, so It may be better to [tell, don't ask](https://www.martinfowler.com/bliki/TellDontAsk.html)

saraycp replied:
Good advice. Can you have a look again, please?

hennevogel commented on src/api/app/models/review.rb:
```diff
@@ -203,11 +203,16 @@ def map_objects_to_ids(objs)
+    params
+  end
+
+  def create_notification(params = {})
```
This is private in `Comment`, why not here?

saraycp replied:
In more recent commits it was renamed to `create_event`. Now I have also moved it to private.

saraycp replied:
BTW, `create_event` should be public in Review model, since it is called from the Request model. Moved back to public...

hennevogel commented on src/api/db/data/20200326170855_remove_obsolete_notifications.rb:
```diff
@@ -0,0 +1,30 @@
+
+  def delete_obsole_notifications
+    Notification.where(notifiable_type: nil, notifiable_id: nil, event_type: EVENT_TYPES).delete_all
+    Notification.where(event_type: AFFECTED_EVENT_TYPES).delete_all
```
Not sure I follow. EVENT_TYPES is old events and AFFECTED_EVENT_TYPES is events that have wrong data?

saraycp replied:
` EVENT_TYPES` means all the events we can create notifications to. I could rename it to `NOTIFIABLE_EVENT_TYPES`.

`AFFECTED_EVENT_TYPES` includes those with wrong data so we have to tackle in this data migration.

The two queries correspond to the two first steps wee planned: 

1. _Remove all Notifications without notifiable_. So I assumed all of the notifiable ones with notifiable set to nil.
2. _Remove all Notifications with `CommentForRequest`, `RequestCreate`, `RequestStatechange`, `ReviewWanted`._ What means the notifiable ones except `CommentForProject` and `CommentForPackage` with notifiable set.

hennevogel replied:
the naming is rather confusing. I would suggest EVENT_TYPES_TO_CHECK_FOR_NOTFIABLE and EVENT_TYPES_TO_REMOVE or something

saraycp replied:
It will be EVENT_TYPES_TO_CHECK_FOR_NOTIFIABLE and  EVENT_TYPES_TO_BE_REMOVED, if are ok.

hennevogel commented on src/api/db/data/20200326170855_remove_obsolete_notifications.rb:
```diff
@@ -0,0 +1,30 @@
+    raise ActiveRecord::IrreversibleMigration
+  end
+
+  def delete_obsole_notifications
```
```suggestion
  def delete_obsolete_notifications
```

saraycp replied:
Done!

hennevogel commented on src/api/spec/db/data/remove_obsolete_notifications_spec.rb:
```diff
@@ -0,0 +1,38 @@
+      expect(Notification.without_notifiable.count).to be_zero
+    end
+
+    it 'deletes all the notifications expect CommentForProject and CommentForPackage' do
```
```suggestion
    it 'deletes all the notifications except CommentForProject and CommentForPackage' do
```

saraycp replied:
Done!

hennevogel commented on src/api/db/data/20200326221616_regenerate_notifications.rb:
```diff
@@ -0,0 +1,98 @@
+    end
+  end
+
+  def new_requests
```
nitpick: Why a method and not a variable? Like

```ruby
new_requests = BsRequest.in_states(:new)
new_requests.each do |request|
...
```

Improves readability IMHO

saraycp replied:
Good advice!

hennevogel commented on src/api/app/models/comment.rb:
```diff
@@ -59,7 +59,7 @@ def notification_parameters(params = {})
 
-  def create_notification(params = {})
+  def create_event(params = {})
     params = notification_parameters(params)
```
`notification_parameters` not?

saraycp replied:
:thinking:  I'm not sure I understand what is wrong with this...

hennevogel replied:
Shouldn't it turn from `create_notification` & `notification_parameters` into `create_event` & `event_parameters`?

saraycp replied:
I've checked and now it is renamed everywhere.

hennevogel commented on src/api/db/data/20200326221616_regenerate_notifications.rb:
```diff
@@ -91,8 +91,11 @@ def recent_request_comments
+      params = subscription.parameters_for_notification.merge!(event.parameters_for_notification)
+      Notification::RssFeedItem.create(params)
+    rescue StandardError => e
+      Airbrake.notify(e, event_id: event.id)
```
What's the difference to "just" let the exception populate to errbit?

saraycp replied:
Let's talk about this.

hennevogel replied:
We talked and came to the conclusion to leave this alone for now

hennevogel commented on src/api/app/services/notification_creator.rb:
```diff
@@ -0,0 +1,27 @@
+    return unless @event.eventtype.in?(EVENTS_TO_NOTIFY)
+    @event.subscriptions.each { |subscription| create_notification_per_subscription(subscription) }
+  rescue StandardError => e
+    Airbrake.notify(e, event_id: @event.id)
```
Same here. what is the difference to let the exception through?

saraycp replied:
Let's talk about this.

hennevogel replied:
We talked and came to the conclusion to leave this alone for now

hennevogel commented:
And 57676c64fbd3963e5b99f6d316dba4c2a8f3492c has the wrong commit message it seems? It talks about review model?

saraycp commented:
@hennevogel I have addressed or answered most of your reviews. 
I also squashed the commits as much as possible and wrote better commit descriptions.
Can you have another look, please?

danidoni commented on src/api/spec/db/data/regenerate_notifications_spec.rb:
```diff
@@ -0,0 +1,183 @@
+    context 'for RequestCreate Notifications' do
+      let!(:subscription) { create(:event_subscription_request_created, receiver_role: 'target_maintainer', user: owner) }
+
+      before do
```
You can use `subject!` above and remove this `before` block, as described [here](https://relishapp.com/rspec/rspec-core/docs/subject/explicit-subject#use-the-%60subject!%60-bang-method-to-call-the-definition-block-before-the-example)

saraycp replied:
As we discussed offline, this is not possible because some `let` are executed after the subject (job) is call so the tests break.

danidoni replied:
Agree

danidoni commented on src/api/spec/services/notification_creator_spec.rb:
```diff
@@ -0,0 +1,39 @@
+        NotificationCreator.new(event).call
+      end
+
+      it 'two CommentForProject notification is created' do
```
Maybe this is the `does not create a duplicated notification` test?

saraycp replied:
Refactor this tests. Can you have another look, please?

danidoni commented on src/api/spec/services/notification_creator_spec.rb:
```diff
@@ -0,0 +1,39 @@
+
+    context 'when the notification already exists' do
+      it 'does not create a duplicated notification' do
+        expect { NotificationCreator.new(event).call }.to change(Notification::RssFeedItem, :count).from(0).to(2)
```
And this test is the `two CommentForProject notification is created`

saraycp replied:
Refactor this tests. Can you have another look, please?

saraycp commented on src/api/db/data/20200326170855_remove_obsolete_notifications.rb:
```diff
@@ -0,0 +1,30 @@
+
+  def delete_obsolete_notifications
+    Notification.where(notifiable_type: nil, notifiable_id: nil, event_type: EVENT_TYPES_TO_CHECK_FOR_NOTIFIABLE).delete_all
+    Notification.where(event_type: EVENT_TYPES_TO_BE_REMOVED).delete_all
```
Good question! I've just pushed another proposal, @hennevogel

hennevogel replied:
Ah finally this makes sense :smile_cat:

krauselukas commented on src/api/app/models/bs_request.rb:
```diff
@@ -1067,6 +1067,11 @@ def required_checks
   end
 
+  # TODO: rename the 'notify_parameters' method instead of wrapping it like this.
+  def event_parameters
```
@saraycp not sure, I dont see the ```event_parameters``` methods used anywhere in the bs_request model, do you wanted to use it in line 782?

krauselukas replied:
besides this the PR LGTM! Great work :)

krauselukas replied:
Sorry forget about it, overlooked something

saraycp commented:
As we talked offline, we decided to split this PR in two.

This PR is the first one and contains the refactoring done in the code related to events and notifications. This PR's description has been updated.

A second PR will be opened soon and  will contain the migration itself.
